scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+        SELECT desiredViews.table_name FROM fileRefTables JOIN desiredViews ON fileRefTables.table_name = desiredViews.table_name
+        """
+
+        return self._hit_bigquery(query)
+
+    def get_duplicates(self, target_table: str) -> Set[str]:
+        """
+        Determines what rows are undesired duplicates. We want to soft delete everything but the latest version for a
+        given entity_id.
+        :param target_table: The particular table to operate on.
+        :return: A set of row ids to soft delete.
+        """
+        sql_table = f"`{self.project}.datarepo_{self.dataset}.{target_table}`"
+
+        # rid -> row_id, fid -> file_id, v -> version
+
+        # allRows:          the row ids, file ids, and versions of all rows in the target table
+        # latestFids:       For all file ids that occur more than once, get the file id and the largest (latest) version
+        # ridsOfAllFids:    The row ids, file ids, and versions of all file ids present in latestFids; in other words,
+        #                   the (row id, file id, version) for every file id that has duplicates
+        # ridsOfLatestFids: The row ids, file ids, and versions of all file ids present in latestFids but ONLY the
+        #                   latest version rows (so a subset of ridsOfAllFids).
+        # Final query:      Get the row ids from ridsOfAllFids but exclude the row ids from ridsOfLatestFids, leaving us
+        #                   with the row ids of all non-latest version file ids. These are the rows to soft delete.
+
+        # Note: The EXCEPT DISTINCT SELECT at the end grabs all row ids of rows that AREN'T the latest version.
+        # The final subquery here is in case there are multiple rows with the same version.
+        query = f"""
+        WITH allRows AS (SELECT datarepo_row_id AS rid, {target_table}_id AS fid, version AS v FROM {sql_table}),
+        latestFids AS (SELECT DISTINCT fid, MAX(v) AS maxv FROM allRows GROUP BY fid HAVING COUNT(1) > 1),
+        ridsOfAllFids AS (SELECT t.rid AS rid, f.fid, t.v FROM latestFids f JOIN allRows t ON t.fid = f.fid),
+        ridsOfLatestFids AS (SELECT t.rid AS rid, f.fid, f.maxv FROM latestFids f JOIN ridsOfAllFids t ON t.fid = f.fid AND t.v = f.maxv)
+        SELECT rid FROM ridsOfAllFids EXCEPT DISTINCT SELECT rid FROM (SELECT MAX(rid) AS rid, fid FROM ridsOfLatestFids GROUP BY fid)
+        """
+
+        return self._hit_bigquery(query)
+
+    def get_null_filerefs(self, target_table: str) -> Set[str]:
+        """
+        Determines what rows have null values in the file_id column. We want to soft delete those.
+        :param target_table: The particular table to operate on.
+        :return: A set of row ids to soft delete.
+        """
+        query = f"""
+        SELECT datarepo_row_id
+        FROM `{self.project}.datarepo_{self.dataset}.{target_table}` WHERE file_id IS NULL
```

Comment:
I usually do SQL things in Python with Postgres and therefore psycopg2 which has modules to help be safer about writing your SQL strings; if there are any suggestions on that front, please help! I don't like writing this stuff out like this where we could very easily muck things up. I guess there's no users besides us so 🤷🏾

Reply:
it's definitely worth diving into the bigquery library you're using to see if it supports parameterization. happy to take a look with you, i have a lot of experience jamming my hand into the source code of poorly documented python APIs.

Reply:
It appears as though BQ and this lib support [parameterization](https://cloud.google.com/bigquery/docs/parameterized-queries), but not for table names.